File: TreeOrderTraversal.java

// Time Complexity : O(log n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : YES
// Any problem you faced while coding this :
// Go thorough recursion logic much more	

// Your code here along with comments explaining your approach

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
     private Map<Integer, List<Integer>> map;
    
    public Solution()
    {
        this.map = new HashMap<>();
    }
    
    public List<List<Integer>> levelOrder(TreeNode root) {
        levelOrderRec(root, 0);
        List<List<Integer>> result = new ArrayList<>();
        
        int index = 0;
        while (map.containsKey(index))
        {
            result.add(map.get(index));
            index++;
        }
        
        return result;
    }
    
    private void levelOrderRec(TreeNode node, int height)
    {
        if (node == null) return;
        
        if (this.map.containsKey(height))
        {
            this.map.get(height).add(node.val);
        } else
        {
            List<Integer> newLevel = new ArrayList<>();
            newLevel.add(node.val);
            this.map.put(height, newLevel);
        }
        
        levelOrderRec(node.left, height + 1);
        levelOrderRec(node.right, height + 1);
    }
}

File: treeorderTraversal.py
from typing import List
import queue
class TreeNode:
    def __init__(self,val=0,left=None,right=None):
        self.val=val
        self.left=left
        self.right=right

class Solution:
    recresult=[]
    def levelOrderqueue(self, root: TreeNode) -> List[List[int]]:
        if root is None:
            return
        qumanuplation = queue.Queue()
        resultstring = ''
        qumanuplation.put(root)
        while not qumanuplation.empty():
            Node = qumanuplation.get()

            resultstring = resultstring + '' + str(Node.val) + ','
            if Node.left:
                qumanuplation.put(Node.left)
            if Node.right:
                qumanuplation.put(Node.right)
        return resultstring



    def levelOrder_quue(self, root: TreeNode) -> List[List[int]]:
        if root is None:
            return
        qumanuplation = queue.Queue()
        result =[]
        qumanuplation.put(root)
        while not qumanuplation.empty():
            size=qumanuplation.qsize()
            temp=[]
            # process level
            for value in range(0,size):
                Node=qumanuplation.get()
                temp.append(Node.val)
                if Node.left:
                    qumanuplation.put(Node.left)
                if Node.right:
                    qumanuplation.put(Node.right)
            result.append(temp)
        return result

    def levelOrder(self, root: TreeNode) -> List[List[int]]:

        if root is None:return root

        self.reccall(root=root,level=0)

        return self.recresult
    def reccall(self,root:TreeNode,level:int) -> None:

        # base
        if root is None:return
        # logic
        if level == len(self.recresult):
            self.recresult.append([])
        self.recresult[level].append(root.val)
        self.reccall(root.left,level+1)
        self.reccall(root.right,level+1)




# time=O(N)
# Space= O(2*N)




if __name__ == '__main__':
    root=TreeNode(4)
    root.right=TreeNode(12)
    root.right.left=TreeNode(11)
    root.right.right=TreeNode(13)
    root.left=TreeNode(2)
    root.left.left=TreeNode(1)
    root.left.right=TreeNode(3)

    print(Solution().levelOrder(root))
